canton/community/participant/src/main/scala/com/digitalasset/canton/participant/admin/PackageService.scala

```diff
@@ -368,22 +369,71 @@ class PackageService(
         .map(_.toRight(s"No such dar ${darId}").flatMap(PackageService.darToLf))
     )
 
-  private def validateArchives(archives: List[DamlLf.Archive])(implicit
+  private def validateArchives(archives: archive.Dar[DamlLf.Archive])(implicit
       traceContext: TraceContext
   ): EitherT[Future, DamlError, Unit] =
     for {
-      packages <- archives
+      mainPackage <- catchUpstreamErrors(Decode.decodeArchive(archives.main))
+      dependencies <- archives.dependencies
         .parTraverse(archive => catchUpstreamErrors(Decode.decodeArchive(archive)))
-        .map(_.toMap)
       _ <- EitherT.fromEither[Future](
         engine
-          .validatePackages(packages)
+          .validatePackages((mainPackage :: dependencies).toMap)
           .leftMap(
             PackageServiceErrors.Validation.handleLfEnginePackageError(_): DamlError
           )
       )
+      _ <- validateUpgrade(mainPackage._1, mainPackage._2)
     } yield ()
 
+  private def validateUpgrade(upgradingPackageId: LfPackageId, upgradingPackage: Package)(implicit
```

Comment:
This is outdated. Can we just backport straight the latest change as of https://github.com/DACH-NY/canton/pull/17662?